Separate Handshake and Notification message interfaces and other tiny changes

kishansagathiya opened the pull request:
## Changes
- Separated Handshake and Notification message interfaces. Removed Hash method from Handshake interface. It was not being used anyway. This would create fewer opportunities where code would get confused between those two. We would not need to cast one interface into other anymore.
- Broke `createNotificationsMessageHandler` into two functions
- Some extra comments

## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues
fixes #2437 
<!-- Write the issue number(s), for example: #123 -->

## Primary Reviewer

<!-- Tag a code owner to review your PR -->

@timwu20

qdm12 commented on dot/network/block_announce.go:
```diff
@@ -18,7 +18,7 @@ import (
 var (
 	_ NotificationsMessage = &BlockAnnounceMessage{}
-	_ NotificationsMessage = &BlockAnnounceHandshake{}
+	_ Handshake            = &BlockAnnounceHandshake{}
```
out of scope nit, but you can change to have zero allocation with

```suggestion
	_ Handshake            = (*BlockAnnounceHandshake)(nil)
```

qdm12 commented on dot/network/block_announce.go:
```diff
@@ -138,20 +138,14 @@ func (*BlockAnnounceHandshake) Type() byte {
-	encMsg, err := hs.Encode()
-	if err != nil {
-		return common.Hash{}, fmt.Errorf("cannot encode handshake: %w", err)
+// IsHandshake returns true if handshakes's role is valid.
+func (hs *BlockAnnounceHandshake) IsHandshake() bool {
```
Since `BlockAnnounceHandshake` is named with `Handshake`, shouldn't it always be a handshake?

kishansagathiya replied:
Renaming this function to `IsValidHandshake` to make things more clear

qdm12 commented on dot/network/notifications.go:
```diff
@@ -128,8 +133,13 @@ func (s *Service) createNotificationsMessageHandler(
+			hs   Handshake
 		)
 
+		if hs, ok = m.(Handshake); ok && hs.IsHandshake() {
```
That feels weird to do an interface assertion and then check if it's a handshake again.

We should do one or the other, not both

timwu20 replied:
Yea do we still need `IsHandshake()` if we can do the type assertions?

kishansagathiya replied:
> That feels weird to do an interface assertion and then check if it's a handshake again.

`IsHandshake` not only checks the type but checks if it is a valid handshake. I am going to rename `IsHandshake` to `IsValidHandshake`, that would make things more clear.

qdm12 commented on dot/network/notifications.go:
```diff
@@ -226,6 +180,56 @@ func (s *Service) createNotificationsMessageHandler(
+	// ie it is an inbound stream and we only send the handshake over it.
+	// we do not send any other data over this stream, we would need to open a new outbound stream.
+	hsData := info.peersData.getInboundHandshakeData(peer)
+	if hsData == nil {
```
nit (a bit out of scope but since these are deltas now) flip the if condition around to return early and have less indentation

qdm12 commented on dot/network/transaction.go:
```diff
@@ -17,7 +17,7 @@ import (
 var (
 	_ NotificationsMessage = &TransactionMessage{}
-	_ NotificationsMessage = &transactionHandshake{}
+	_ Handshake            = &transactionHandshake{}
```
nit

```suggestion
	_ Handshake            = (*transactionHandshake)(nil)
```

qdm12 commented on dot/network/notifications.go:
```diff
@@ -226,6 +179,58 @@ func (s *Service) createNotificationsMessageHandler(
+	// once validated, send back a handshake
+	resp, err := info.getHandshake()
+	if err != nil {
+		logger.Warnf("failed to get handshake using protocol %s: %s", info.protocolID, err)
```
nit I know it's copy pasta, but since it shows as deltas, one more suggestion...

Why is this logged at the warn level whereas others are logged at the trace level? Would it feel appropriate to remove this log and rely on the caller Trace log instead?



```suggestion
```

kishansagathiya replied:
> Would it feel appropriate to remove this log and rely on the caller Trace log instead?

I prefer not to ignore errors. If we are not returning the err, at least we should log it as warn.

qdm12 reviewed (commented):
Nice changes 👍 👍 

Just a few nits (most should just be easy-to-batch-suggestion-commit) about the copied pasted code since it shows as deltas.

EclesioMeloJunior commented on lib/grandpa/network.go:
```diff
@@ -63,8 +63,8 @@ func (*GrandpaHandshake) Hash() (common.Hash, error) {
-// IsHandshake returns true
-func (*GrandpaHandshake) IsHandshake() bool {
+// IsValidHandshake returns true
+func (*GrandpaHandshake) IsValidHandshake() bool {
```
would be nice to have a similar check here since the `GrandpaHandshake` should come from other authority peers
```go
switch hs.Roles {
	case common.AuthorityRole:
		return true
	default:
		return false
	}
```

kishansagathiya replied:
addressed

EclesioMeloJunior reviewed (changes requested):
LGTM, just a small adjustment

timwu20 commented on dot/network/notifications.go:
```diff
@@ -20,7 +20,9 @@ const handshakeTimeout = time.Second * 10
-	NotificationsMessage
+	Message
+	Type() byte
+	IsValidHandshake() bool
```
do we even need an interface for this anymore? Could we not do type checks on the types that are handshake messages? Or even use the constraint feature in go generics.

kishansagathiya replied:
`IsValidHandshake` validates the handshake as well. It does more that type checking.

timwu20 commented on dot/network/message.go:
```diff
@@ -36,7 +36,7 @@ type NotificationsMessage interface {
 	Type() byte
 	Hash() (common.Hash, error)
-	IsHandshake() bool
+	IsValidHandshake() bool
```
Do we still need an interface for this? Could this be refactored further using generics? As it stands, all handshake messages implement every method of this interface except `Hash()`.

timwu20 commented on dot/network/block_announce.go:
```diff
@@ -18,7 +18,7 @@ import (
 var (
 	_ NotificationsMessage = &BlockAnnounceMessage{}
-	_ NotificationsMessage = &BlockAnnounceHandshake{}
+	_ Handshake            = (*BlockAnnounceHandshake)(nil)
```
why the nil case?

kishansagathiya replied:
This is a check to make sure that BlockAnnounceHandshake implements Handshake interface. I think, Unlike the previous times, with nil, we end up not allocating any memory.

timwu20 commented on dot/network/transaction.go:
```diff
@@ -17,7 +17,7 @@ import (
 var (
 	_ NotificationsMessage = &TransactionMessage{}
-	_ NotificationsMessage = &transactionHandshake{}
+	_ Handshake            = (*transactionHandshake)(nil)
```
why the nil case?

kishansagathiya replied:
This is a check to make sure that transactionHandshake implements Handshake interface. I think, Unlike the previous times, with nil, we end up not allocating any memory.

kishansagathiya commented:
@timwu20 There was a opportunity to do more refactor. Removed IsValidHandshake() from NotificationsMessage Interface and removed Type() from Handshake interface. Bother interfaces are different now (can't be casted into each other)
